ext/telemetry/telemetry.ts

```diff
@@ -687,260 +700,515 @@ class ContextManager {
-                );
-                j += 3;
-              } else if (j + 2 < boundaries.length) {
-                op_otel_metrics_histogram_data_point_entry2(
-                  counts[j],
-                  boundaries[j],
-                  counts[j + 1],
-                  boundaries[j + 1],
-                );
-                j += 2;
-              } else {
-                op_otel_metrics_histogram_data_point_entry1(
-                  counts[j],
-                  boundaries[j],
-                );
-                j += 1;
-              }
-            }
-            op_otel_metrics_histogram_data_point_entry_final(counts[j]);
-            const attributes = ObjectEntries(dataPoint.attributes);
-            let i = 0;
-            while (i < attributes.length) {
-              if (i + 2 < attributes.length) {
-                op_otel_metrics_data_point_attribute3(
-                  attributes.length,
-                  attributes[i][0],
-                  attributes[i][1],
-                  attributes[i + 1][0],
-                  attributes[i + 1][1],
-                  attributes[i + 2][0],
-                  attributes[i + 2][1],
-                );
-                i += 3;
-              } else if (i + 1 < attributes.length) {
-                op_otel_metrics_data_point_attribute2(
-                  attributes.length,
-                  attributes[i][0],
-                  attributes[i][1],
-                  attributes[i + 1][0],
-                  attributes[i + 1][1],
-                );
-                i += 2;
-              } else {
-                op_otel_metrics_data_point_attribute(
-                  attributes.length,
-                  attributes[i][0],
-                  attributes[i][1],
-                );
-                i += 1;
-              }
-            }
-          }
-          break;
-        default:
-          continue;
+  addBatchObservableCallback(
+    callback: BatchObservableCallback,
+    observables: Observable[],
+  ): void {
+    if (!METRICS_ENABLED) return;
+    const result = new BatchObservableResult(new SafeWeakSet(observables));
+    startObserving();
+    BATCH_CALLBACKS.set(callback, result);
+  }
+
+  removeBatchObservableCallback(
+    callback: BatchObservableCallback,
+    observables: Observable[],
+  ): void {
+    if (!METRICS_ENABLED) return;
+    const result = BATCH_CALLBACKS.get(callback);
+    if (result && batchResultHasObservables(result, observables)) {
+      BATCH_CALLBACKS.delete(callback);
+    }
+  }
+}
+
+class NoopInstrument {
+  add(_value: number, _attributes?: MetricAttributes, _context?: Context) {}
+  record(_value: number, _attributes?: MetricAttributes, _context?: Context) {}
+  addCallback(_callback: ObservableCallback) {}
+  removeCallback(_callback: ObservableCallback) {}
+}
+
+type BatchObservableCallback = (
+  observableResult: BatchObservableResult,
+) => void | Promise<void>;
+
+function record(
+  instrument: Instrument,
```

Comment:
if this accepts `Instrument|null` you can pipe that through the normal classes based on `METRICS_ENABLED` without needing a NoopInstrument class (smth like `let instrument = METRICS_ENABLED ? op_create_instrument() : null`)